Solved incorrect left limit result for atan at inf

borgesaugusto opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #25582 
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
As suggested by [anutosh491](https://github.com/anutosh491) PoleError is raised in case the value is nan, which solves the bug. A test was added to check for this case.


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* functions
   * Solved bug with atan when solving a left inf limit 
<!-- END RELEASE NOTES -->

borgesaugusto committed:
author: add Augusto Borges to .mailmap

borgesaugusto committed:
trigonometric: Solves incorrect inf limit for atan

smichr commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -2677,6 +2677,8 @@ def taylor_term(n, x, *previous_terms):
     def _eval_as_leading_term(self, x, logx=None, cdir=0):  # atan
         arg = self.args[0]
         x0 = arg.subs(x, 0).cancel()
+        if x0 is S.NaN:
```
The other 2 times that PoleError is raised in this file there is a message in the error that tells what failed. Can a similar message be added in this case? What specifically just happened that caused the PoleError to raise?

anutosh491 commented on sympy/functions/elementary/tests/test_trigonometric.py:
```diff
@@ -2160,3 +2160,7 @@ def test_as_real_imag():
     assert e3.period() == 8
+
+
+def test_atan_inf_left():
```
I do believe that we would like to see some sort of `PoleError` here but I would like to test out more cases (think of cases that could go wrong) . the `acot` function is related to the `atan` function hence there is a possiblity of cases going wrong there too.
```
>>> atan(x).rewrite(acot)
acot(1/x)
```

Also could you please change the name of the function (use `test_issue_issue_number` like other functions) and place it in sorted order

anutosh491 commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -2677,6 +2677,8 @@ def taylor_term(n, x, *previous_terms):
         x0 = arg.subs(x, 0).cancel()
+        if x0 is S.NaN:
+            raise PoleError()
         if x0.is_zero:
```
Also another thing being we would encounter this case in all inverse trigonometric functions (while calculating limits for 
`asin(exp(x))`, `acsc(exp(x))`). Just addressing an individual case might not help us overall, could you try tackling all cases that go wrong.

jksuom replied:
`nan` typically results when a substitution gives 0/0 (e.g. `arg = sin(x)/x`). It may be advisable to first try to compute `x0` as a limit before raising `PoleError`.

smichr replied:
e.g. `limit(acos(sin(x)/x),x,oo)` ? @borgesaugusto, what does that do in this PR?

borgesaugusto replied:
This: limit(acos(sin(x)),x,oo) I*log(I*AccumBounds(-1, 1) + AccumBounds(0, 1)) + pi/2
and 
limit(acos(sin(x)/x),x,oo) gives "sympy.core.function.PoleError: Cannot expand sin(1/_w) around 0" in "sympy/functions/elementary/trigonometric.py" line 436" (The txt with the trace), ending with a NotImplementedError.
[sympy_trace.txt](https://github.com/sympy/sympy/files/12840213/sympy_trace.txt)

borgesaugusto committed:
Merge branch 'sympy:master' into iss_25582

borgesaugusto committed:
Test moved to test_limits. All inverse trigonometric functions check for nan

borgesaugusto commented:
I added the check to all the inverse trigonometric functions (except atan2) and moved the test to the suggested file (incorporating the asserts for the new functions and changing the name). 
I don't see a more general way of doing it.

smichr commented:
This seems like a positive step forward. Thanks, this is in.
